pulsar-client-cpp/lib/Future.h

```diff
@@ -99,21 +100,24 @@ class Promise {
         }
 
         state->value = value;
-        state->result = Result();
+        state->result = DEFAULT_RESULT;
         state->complete = true;
 
-        typename std::list<ListenerCallback>::iterator it;
-        for (it = state->listeners.begin(); it != state->listeners.end(); ++it) {
-            ListenerCallback& callback = *it;
-            callback(state->result, state->value);
+        const auto listeners = std::move(state->listeners);
+
+        lock.unlock();
+
+        for (auto& callback : listeners) {
+            callback(DEFAULT_RESULT, value);
         }
 
         state->listeners.clear();
```

Comment:
the `state->listeners.clear()` would have to be done from within the mutex. Probably it might not be needed anymore? In what state is the `std::vector` left after the listeners are moved out?

Reply:
It's right. I just forgot to remove it. (I have removed it from `setFailed`)

Reply:
The only thing I’m not 100 sure if it’s safe to keep using the listener vector after it's moved out. Eg. If a new listener is added after the future  is completed, it will add again into the moved vector which is in an undefined state. A I understand, the only guarantee is that the vector will be empty, but we have no idea whether we can add or not. 

Instead of the move, it might actually be safer to do a swap with an empty std::vector.

Reply:
Sounds reasonable. Ideally we should not use the object that has been moved. I'll adopt the swap way.